HIP/hip-1195.md

```diff
@@ -0,0 +1,834 @@
+  // ...
+
+  /**
+   * The hooks to install on the account.
+   */
+  repeated HookInstall hook_installs = 19;
+
+  /**
+   * The indexes of the hooks to uninstall from the account.
+   */
+  repeated uint64 hook_indexes_to_uninstall = 20;
+}
+```
+
+The `Account` message in `TokenService` state is extended to include the number of installed hooks, as well as
+the indexes of the first last hooks in the doubly-linked list of hooks installed by the account.
+
+```protobuf
+message Account {
+  // ...
+  /**
+   * The number of hook currently installed on this account.
+   */
+  uint64 number_installed_hooks = 36;
+
+  /**
+   * If positive, the index of the first hook installed on this account.
+   */
+  uint64 first_hook_index = 37;
+}
+```
+
+For a successful such `CryptoCreate` or `CryptoUpdate`, the indexes of the newly installed hooks will appear in the
+legacy record `TransactionReceipt` if record streams are still enabled.
+
+Now we need to let a `CryptoTransfer` reference such a hook. For this we extend the `AccountAmount` and `NftTransfer`
+messages used in the `CryptoTransferTransactionBody`.
+
+```protobuf
+message AccountAmount {
+  // ...
+  /**
+   * If set, a call to a hook of type `ACCOUNT_ALLOWANCE_HOOK` installed on
+   * accountID that must succeed for the transaction to occur.
+   */
+  HookCall allowance_hook_call = 4;
+}
+
+message NftTransfer {
+  // ...
+  /**
+   * If set, a call to a hook of type `ACCOUNT_ALLOWANCE_HOOK` installed on
+   * senderAccountID that must succeed for the transaction to occur.
+   */
+  HookCall sender_allowance_hook_call = 5;
+
+  /**
+   * If set, a call to a hook of type `ACCOUNT_ALLOWANCE_HOOK` installed on
+   * receiverAccountID that must succeed for the transaction to occur.
+   */
+  HookCall receiver_allowance_hook_call = 6;
+}
+```
+
+Note that `NftTransfer` supports both sender and receiver transfer allowance hooks, since the transaction may
+need to use the receiver hook to satisfy a `receiver_sig_required=true` setting.
+
+### The transfer allowance ABI
+
+The account allowance EVM hook ABI is as follows,
+
+```solidity
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity >=0.4.9 <0.9.0;
+pragma experimental ABIEncoderV2;
+
+import './IHederaTokenService.sol';
+
+/// The interface for an account allowance EVM hook.
+interface IHieroAccountAllowanceEvmHook {
+    /// Combines HBAR and HTS asset transfers.
+    struct Transfers {
+        /// The HBAR transfers
+        IHederaTokenService.TransferList hbar;
+        /// The HTS token transfers
+        IHederaTokenService.TokenTransferList[] tokens;
+    }
+
+    /// Combines the full proposed transfers for a Hiero transaction,
+    /// including both its direct transfers and the implied HIP-18
+    /// custom fee transfers.
+    struct ProposedTransfers {
+        /// The transaction's direct transfers
+        Transfers direct;
+        /// The transaction's assessed custom fees
+        Transfers customFee;
+    }
```

Comment:
In discussions with THF and other DAs, we found that many use cases could benefit from knowledge of the **memo field** and **network fees** associated to the transaction when deciding whether to accept a transaction.

I suggest renaming this structure to something more generic, such as `TransactionContext', and including both the above pieces of information. Or just to create a new struct and add that struct to the `allow` function parameters.